Firestore/Source/Model/FSTFieldValue.mm

```diff
@@ -1027,4 +959,114 @@ - (NSComparisonResult)compare:(FSTFieldValue *)other {
+@property(nonatomic, assign, readonly) FieldValue internalValue;
+@end
+
+@implementation FSTDelegateValue
++ (instancetype)delegateValue:(FieldValue &&)value {
+  return [[FSTDelegateValue alloc] initWithValue:std::move(value)];
+}
+
+- (id)initWithValue:(FieldValue &&)value {
+  self = [super init];
+  if (self) {
+    _internalValue = std::move(value);
+  }
+  return self;
+}
+
+- (FieldValue::Type)type {
+  return self.internalValue.type();
+}
+
+- (FSTTypeOrder)typeOrder {
+  switch (self.internalValue.type()) {
+    case FieldValue::Type::Null:
+      return FSTTypeOrderNull;
+    case FieldValue::Type::Boolean:
+      return FSTTypeOrderBoolean;
+    case FieldValue::Type::Integer:
+    case FieldValue::Type::Double:
+      return FSTTypeOrderNumber;
+    case FieldValue::Type::Timestamp:
+    case FieldValue::Type::ServerTimestamp:
+      return FSTTypeOrderTimestamp;
+    case FieldValue::Type::String:
+      return FSTTypeOrderString;
+    case FieldValue::Type::Blob:
+      return FSTTypeOrderBlob;
+    case FieldValue::Type::Reference:
+      return FSTTypeOrderReference;
+    case FieldValue::Type::GeoPoint:
+      return FSTTypeOrderGeoPoint;
+    case FieldValue::Type::Array:
+      return FSTTypeOrderArray;
+    case FieldValue::Type::Object:
+      return FSTTypeOrderObject;
+  }
+  UNREACHABLE();
+}
+
+- (BOOL)isEqual:(id)other {
+  // Simplification: We'll assume that (eg) FSTBooleanValue(true) !=
+  // FSTDelegateValue(FieldValue::FromBoolean(true)). That's not great. We'll
+  // handle this by ensuring that we remove (eg) FSTBooleanValue at the same
+  // time that FSTDelegateValue handles (eg) booleans to ensure this case never
+  // occurs.
+
+  if (other == self) {
+    return YES;
+  }
+  if (![other isKindOfClass:[self class]]) {
+    return NO;
+  }
+
+  return self.internalValue == ((FSTDelegateValue *)other).internalValue;
+}
+
+- (id)value {
+  switch (self.internalValue.type()) {
+    case FieldValue::Type::Null:
+      HARD_FAIL("TODO(rsgowman): implement");
+    case FieldValue::Type::Boolean:
+      return self.internalValue.boolean_value() ? @YES : @NO;
+    case FieldValue::Type::Integer:
+    case FieldValue::Type::Double:
+    case FieldValue::Type::Timestamp:
+    case FieldValue::Type::ServerTimestamp:
+    case FieldValue::Type::String:
+    case FieldValue::Type::Blob:
+    case FieldValue::Type::Reference:
+    case FieldValue::Type::GeoPoint:
+    case FieldValue::Type::Array:
+    case FieldValue::Type::Object:
+      HARD_FAIL("TODO(rsgowman): implement");
+  }
+  UNREACHABLE();
+}
+
+- (NSComparisonResult)compare:(FSTFieldValue *)other {
+  // Simplification: We'll assume that if self.type == other.type, then other
+  // must be a FSTDelegateValue. That's not great. We'll handle this by
+  // ensuring that we remove (eg) FSTBooleanValue at the same time that
+  // FSTDelegateValue handles (eg) booleans to ensure this case never occurs.
+
+  if (self.type == other.type) {
```

Comment:
This should be `FieldValue::Comarable(self.type, other.type)`. There are cases where the types aren't equal but comparison should happen (e.g. timestamp types and numeric types).